Python-driven testing for the new operator registration API

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35058 C++ smoketests for the API.
* **#34668 Python-driven testing for the new operator registration API**
* #34620 Minor OperatorName cleanup
* #34618 Improved error checking
* #34601 Implement schema matching inside Dispatcher
* #34456 Finish implementing schema-less impl, switch all internal registrations to it
* #34455 Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)
* #34221 Relax internal invariants of Dispatcher in preparation for schema-less registration

The general concept:
- Bind a (very gimped) version of the dispatcher API from Python,
  so that we can easily write a more complex testing harness
  using expect tests.
- For series of registrations we want to test, exhaustively
  test every possible permutation of registrations (and
  deregistrations), and show that the intermediate states
  agree no matter what path is taken.
- Intermediate states are rendered using a new dumpState()
  debugging method that prints the internal state of the
  dispatcher.  This method may be generally useful for people
  who want to see what's in the dispatcher.
- Simultaneously, add a new invariant testing function which
  checks that the internal invariants of the dispatcher are
  upheld (so we don't have to print internal implementation
  details of the dispatcher)

Some new features we added to make the testing possible:
- Registered kernel functions now store a debug string which
  can be used to more easily identify them.  There's some best
  effort stuff based on __FUNCSIG__ but this is only really
  capable of reporting types and not function symbols.  Tests
  use this to distinguish between multiple distinct registrations.
- Rename KernelSchemaPair back to KernelEntry (so it accurately
  stores debug string too)
- Refactored the op registration API to take FunctionSchema directly.
  We now do namespacing by post facto fixing up the OperatorName
  embedded in FunctionSchema.  This also means that you can
  now do torch::import("ns1").def("ns2::blah") and have the ns2
  override ns1 (although maybe this is not the correct behavior.)
- New torch::schema public API, for attaching alias analysis kind
  annotation kinds.  This meant we had to template up some function
  signatures which previously took const char*.  There's now a nice
  comment explaining this strategy.
- torch::import now takes std::string which means we can use
  the namespacing from Python
- BC shim for subTest; I ended up not using this in the end

Fixed a bug where we register a schema too early before doing error
checking.  This was a nice test for the testing framework, which
found the error:

```
Traceback (most recent call last):
  File "test/test_dispatch.py", line 164, in test_def_impl_schema_mismatch
    ], raises=True)
  File "test/test_dispatch.py", line 135, in commute
    results=results, raises=raises)
  File "test/test_dispatch.py", line 83, in run_permutation
    .format(ctor_order[:i], op_ix))
  File "test/test_dispatch.py", line 59, in check_invariants
    .format(expected_provenance, actual_provenance)
AssertionError: 'name[16 chars]ema: (none)\ncatchall: boxed unboxed :: (Tenso[18 chars]0)\n' != 'name[16 chars]ema: test::foo(Tensor x, Tensor y) -> (Tensor)[53 chars]0)\n'
  name: test::foo
- schema: (none)
+ schema: test::foo(Tensor x, Tensor y) -> (Tensor)
  catchall: boxed unboxed :: (Tensor _0) -> (Tensor _0)
 : expected from running ctors (1,); actual from running ctors (1,) and then failing to run ctor 0 (did this failure leave the dispatcher in a wedged state? it shouldn't!)
```

Coming soon: direct C++ tests for basic API surface area

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
WIP

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "WIP"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "WIP"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "WIP"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Python-driven testing for the new operator registration API"


The general concept:
- Bind a (very gimped) version of the dispatcher API from Python,
  so that we can easily write a more complex testing harness
  using expect tests.
- For series of registrations we want to test, exhaustively
  test every possible permutation of registrations (and
  deregistrations), and show that the intermediate states
  agree no matter what path is taken.
- Intermediate states are rendered using a new dumpState()
  debugging method that prints the internal state of the
  dispatcher.  This method may be generally useful for people
  who want to see what's in the dispatcher.
- Simultaneously, add a new invariant testing function which
  checks that the internal invariants of the dispatcher are
  upheld (so we don't have to print internal implementation
  details of the dispatcher)

Some new features we added to make the testing possible:
- Registered kernel functions now store a debug string which
  can be used to more easily identify them.  There's some best
  effort stuff based on __FUNCSIG__ but this is only really
  capable of reporting types and not function symbols.  Tests
  use this to distinguish between multiple distinct registrations.
- Rename KernelSchemaPair back to KernelEntry (so it accurately
  stores debug string too)
- Refactored the op registration API to take FunctionSchema directly.
  We now do namespacing by post facto fixing up the OperatorName
  embedded in FunctionSchema.  This also means that you can
  now do torch::import("ns1").def("ns2::blah") and have the ns2
  override ns1 (although maybe this is not the correct behavior.)
- New torch::schema public API, for attaching alias analysis kind
  annotation kinds.  This meant we had to template up some function
  signatures which previously took const char*.  There's now a nice
  comment explaining this strategy.
- torch::import now takes std::string which means we can use
  the namespacing from Python
- BC shim for subTest; I ended up not using this in the end

Fixed a bug where we register a schema too early before doing error
checking.  This was a nice test for the testing framework, which
found the error:

```
Traceback (most recent call last):
  File "test/test_dispatch.py", line 164, in test_def_impl_schema_mismatch
    ], raises=True)
  File "test/test_dispatch.py", line 135, in commute
    results=results, raises=raises)
  File "test/test_dispatch.py", line 83, in run_permutation
    .format(ctor_order[:i], op_ix))
  File "test/test_dispatch.py", line 59, in check_invariants
    .format(expected_provenance, actual_provenance)
AssertionError: 'name[16 chars]ema: (none)\ncatchall: boxed unboxed :: (Tenso[18 chars]0)\n' != 'name[16 chars]ema: test::foo(Tensor x, Tensor y) -> (Tensor)[53 chars]0)\n'
  name: test::foo
- schema: (none)
+ schema: test::foo(Tensor x, Tensor y) -> (Tensor)
  catchall: boxed unboxed :: (Tensor _0) -> (Tensor _0)
 : expected from running ctors (1,); actual from running ctors (1,) and then failing to run ctor 0 (did this failure leave the dispatcher in a wedged state? it shouldn't!)
```

Coming soon: direct C++ tests for basic API surface area

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Python-driven testing for the new operator registration API"


The general concept:
- Bind a (very gimped) version of the dispatcher API from Python,
  so that we can easily write a more complex testing harness
  using expect tests.
- For series of registrations we want to test, exhaustively
  test every possible permutation of registrations (and
  deregistrations), and show that the intermediate states
  agree no matter what path is taken.
- Intermediate states are rendered using a new dumpState()
  debugging method that prints the internal state of the
  dispatcher.  This method may be generally useful for people
  who want to see what's in the dispatcher.
- Simultaneously, add a new invariant testing function which
  checks that the internal invariants of the dispatcher are
  upheld (so we don't have to print internal implementation
  details of the dispatcher)

Some new features we added to make the testing possible:
- Registered kernel functions now store a debug string which
  can be used to more easily identify them.  There's some best
  effort stuff based on __FUNCSIG__ but this is only really
  capable of reporting types and not function symbols.  Tests
  use this to distinguish between multiple distinct registrations.
- Rename KernelSchemaPair back to KernelEntry (so it accurately
  stores debug string too)
- Refactored the op registration API to take FunctionSchema directly.
  We now do namespacing by post facto fixing up the OperatorName
  embedded in FunctionSchema.  This also means that you can
  now do torch::import("ns1").def("ns2::blah") and have the ns2
  override ns1 (although maybe this is not the correct behavior.)
- New torch::schema public API, for attaching alias analysis kind
  annotation kinds.  This meant we had to template up some function
  signatures which previously took const char*.  There's now a nice
  comment explaining this strategy.
- torch::import now takes std::string which means we can use
  the namespacing from Python
- BC shim for subTest; I ended up not using this in the end

Fixed a bug where we register a schema too early before doing error
checking.  This was a nice test for the testing framework, which
found the error:

```
Traceback (most recent call last):
  File "test/test_dispatch.py", line 164, in test_def_impl_schema_mismatch
    ], raises=True)
  File "test/test_dispatch.py", line 135, in commute
    results=results, raises=raises)
  File "test/test_dispatch.py", line 83, in run_permutation
    .format(ctor_order[:i], op_ix))
  File "test/test_dispatch.py", line 59, in check_invariants
    .format(expected_provenance, actual_provenance)
AssertionError: 'name[16 chars]ema: (none)\ncatchall: boxed unboxed :: (Tenso[18 chars]0)\n' != 'name[16 chars]ema: test::foo(Tensor x, Tensor y) -> (Tensor)[53 chars]0)\n'
  name: test::foo
- schema: (none)
+ schema: test::foo(Tensor x, Tensor y) -> (Tensor)
  catchall: boxed unboxed :: (Tensor _0) -> (Tensor _0)
 : expected from running ctors (1,); actual from running ctors (1,) and then failing to run ctor 0 (did this failure leave the dispatcher in a wedged state? it shouldn't!)
```

Coming soon: direct C++ tests for basic API surface area

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bhosmer reviewed (commented):
Nice! Few minor nits/questions but looks solid to me. Cool that it's already paying its way too :)

Leaving the final signoff to @smessmer.

ezyang committed:
Update on "Python-driven testing for the new operator registration API"


The general concept:
- Bind a (very gimped) version of the dispatcher API from Python,
  so that we can easily write a more complex testing harness
  using expect tests.
- For series of registrations we want to test, exhaustively
  test every possible permutation of registrations (and
  deregistrations), and show that the intermediate states
  agree no matter what path is taken.
- Intermediate states are rendered using a new dumpState()
  debugging method that prints the internal state of the
  dispatcher.  This method may be generally useful for people
  who want to see what's in the dispatcher.
- Simultaneously, add a new invariant testing function which
  checks that the internal invariants of the dispatcher are
  upheld (so we don't have to print internal implementation
  details of the dispatcher)

Some new features we added to make the testing possible:
- Registered kernel functions now store a debug string which
  can be used to more easily identify them.  There's some best
  effort stuff based on __FUNCSIG__ but this is only really
  capable of reporting types and not function symbols.  Tests
  use this to distinguish between multiple distinct registrations.
- Rename KernelSchemaPair back to KernelEntry (so it accurately
  stores debug string too)
- Refactored the op registration API to take FunctionSchema directly.
  We now do namespacing by post facto fixing up the OperatorName
  embedded in FunctionSchema.  This also means that you can
  now do torch::import("ns1").def("ns2::blah") and have the ns2
  override ns1 (although maybe this is not the correct behavior.)
- New torch::schema public API, for attaching alias analysis kind
  annotation kinds.  This meant we had to template up some function
  signatures which previously took const char*.  There's now a nice
  comment explaining this strategy.
- torch::import now takes std::string which means we can use
  the namespacing from Python
- BC shim for subTest; I ended up not using this in the end

Fixed a bug where we register a schema too early before doing error
checking.  This was a nice test for the testing framework, which
found the error:

```
Traceback (most recent call last):
  File "test/test_dispatch.py", line 164, in test_def_impl_schema_mismatch
    ], raises=True)
  File "test/test_dispatch.py", line 135, in commute
    results=results, raises=raises)
  File "test/test_dispatch.py", line 83, in run_permutation
    .format(ctor_order[:i], op_ix))
  File "test/test_dispatch.py", line 59, in check_invariants
    .format(expected_provenance, actual_provenance)
AssertionError: 'name[16 chars]ema: (none)\ncatchall: boxed unboxed :: (Tenso[18 chars]0)\n' != 'name[16 chars]ema: test::foo(Tensor x, Tensor y) -> (Tensor)[53 chars]0)\n'
  name: test::foo
- schema: (none)
+ schema: test::foo(Tensor x, Tensor y) -> (Tensor)
  catchall: boxed unboxed :: (Tensor _0) -> (Tensor _0)
 : expected from running ctors (1,); actual from running ctors (1,) and then failing to run ctor 0 (did this failure leave the dispatcher in a wedged state? it shouldn't!)
```

Coming soon: direct C++ tests for basic API surface area

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.cpp:
```diff
@@ -16,4 +18,27 @@ void fallthrough_kernel(OperatorKernel*, const OperatorHandle&, Stack*) {
+// single line summary of state
+std::string KernelFunction::dumpState() const {
+  std::ostringstream oss;
+  if (boxed_kernel_func_ == fallthrough_kernel) {
```
sure, why not

ezyang commented on torch/csrc/utils/python_dispatch.cpp:
```diff
@@ -0,0 +1,162 @@
+  };
+  auto it = key_map.find(k);
+  TORCH_CHECK(it != key_map.end(), "could not parse ", k);
+  if (it->second == c10::DispatchKey::Undefined) {
```
It's because I'm a lazy bastard. If I directly store nullopt in the map, then I have to `make_optional` every other entry in the map (to get the types to line up).

Undefined already exists as the "nullopt" element for DispatchKey. For example, this existing comment:

```
  // ~~~~~~~~~~~~~~~~~~~~~~~~~~ UNDEFINED ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ //
  // This is not a "real" tensor id, but it exists to give us a "nullopt"
  // element we can return for cases when a DispatchKeySet contains no elements.
  // You can think a more semantically accurate definition of DispatchKey is:
  //
  //    using DispatchKey = optional<RealDispatchKey>
  //
  // and Undefined == nullopt.  We didn't actually represent
  // it this way because optional<RealDispatchKey> would take two
  // words, when DispatchKey fits in eight bits.
```

So, technically, when you say `optional<DispatchKey>`, this is actually `optional<optional<RealDispatchKey>>`, which is rarely what you want. But we have a similar situation with Tensor, which also has a nullopt element (the default constructor), but we still opt to say `optional<Tensor>` in some cases to be more explicit about the null-ness (and the default constructor is illegal here). That's kind of what's going on here.

ezyang committed:
Update on "Python-driven testing for the new operator registration API"


The general concept:
- Bind a (very gimped) version of the dispatcher API from Python,
  so that we can easily write a more complex testing harness
  using expect tests.
- For series of registrations we want to test, exhaustively
  test every possible permutation of registrations (and
  deregistrations), and show that the intermediate states
  agree no matter what path is taken.
- Intermediate states are rendered using a new dumpState()
  debugging method that prints the internal state of the
  dispatcher.  This method may be generally useful for people
  who want to see what's in the dispatcher.
- Simultaneously, add a new invariant testing function which
  checks that the internal invariants of the dispatcher are
  upheld (so we don't have to print internal implementation
  details of the dispatcher)

Some new features we added to make the testing possible:
- Registered kernel functions now store a debug string which
  can be used to more easily identify them.  There's some best
  effort stuff based on __FUNCSIG__ but this is only really
  capable of reporting types and not function symbols.  Tests
  use this to distinguish between multiple distinct registrations.
- Rename KernelSchemaPair back to KernelEntry (so it accurately
  stores debug string too)
- Refactored the op registration API to take FunctionSchema directly.
  We now do namespacing by post facto fixing up the OperatorName
  embedded in FunctionSchema.  This also means that you can
  now do torch::import("ns1").def("ns2::blah") and have the ns2
  override ns1 (although maybe this is not the correct behavior.)
- New torch::schema public API, for attaching alias analysis kind
  annotation kinds.  This meant we had to template up some function
  signatures which previously took const char*.  There's now a nice
  comment explaining this strategy.
- torch::import now takes std::string which means we can use
  the namespacing from Python
- BC shim for subTest; I ended up not using this in the end

Fixed a bug where we register a schema too early before doing error
checking.  This was a nice test for the testing framework, which
found the error:

```
Traceback (most recent call last):
  File "test/test_dispatch.py", line 164, in test_def_impl_schema_mismatch
    ], raises=True)
  File "test/test_dispatch.py", line 135, in commute
    results=results, raises=raises)
  File "test/test_dispatch.py", line 83, in run_permutation
    .format(ctor_order[:i], op_ix))
  File "test/test_dispatch.py", line 59, in check_invariants
    .format(expected_provenance, actual_provenance)
AssertionError: 'name[16 chars]ema: (none)\ncatchall: boxed unboxed :: (Tenso[18 chars]0)\n' != 'name[16 chars]ema: test::foo(Tensor x, Tensor y) -> (Tensor)[53 chars]0)\n'
  name: test::foo
- schema: (none)
+ schema: test::foo(Tensor x, Tensor y) -> (Tensor)
  catchall: boxed unboxed :: (Tensor _0) -> (Tensor _0)
 : expected from running ctors (1,); actual from running ctors (1,) and then failing to run ctor 0 (did this failure leave the dispatcher in a wedged state? it shouldn't!)
```

Coming soon: direct C++ tests for basic API surface area

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Python-driven testing for the new operator registration API"


The general concept:
- Bind a (very gimped) version of the dispatcher API from Python,
  so that we can easily write a more complex testing harness
  using expect tests.
- For series of registrations we want to test, exhaustively
  test every possible permutation of registrations (and
  deregistrations), and show that the intermediate states
  agree no matter what path is taken.
- Intermediate states are rendered using a new dumpState()
  debugging method that prints the internal state of the
  dispatcher.  This method may be generally useful for people
  who want to see what's in the dispatcher.
- Simultaneously, add a new invariant testing function which
  checks that the internal invariants of the dispatcher are
  upheld (so we don't have to print internal implementation
  details of the dispatcher)

Some new features we added to make the testing possible:
- Registered kernel functions now store a debug string which
  can be used to more easily identify them.  There's some best
  effort stuff based on __FUNCSIG__ but this is only really
  capable of reporting types and not function symbols.  Tests
  use this to distinguish between multiple distinct registrations.
- Rename KernelSchemaPair back to KernelEntry (so it accurately
  stores debug string too)
- Refactored the op registration API to take FunctionSchema directly.
  We now do namespacing by post facto fixing up the OperatorName
  embedded in FunctionSchema.  This also means that you can
  now do torch::import("ns1").def("ns2::blah") and have the ns2
  override ns1 (although maybe this is not the correct behavior.)
- New torch::schema public API, for attaching alias analysis kind
  annotation kinds.  This meant we had to template up some function
  signatures which previously took const char*.  There's now a nice
  comment explaining this strategy.
- torch::import now takes std::string which means we can use
  the namespacing from Python
- BC shim for subTest; I ended up not using this in the end

Fixed a bug where we register a schema too early before doing error
checking.  This was a nice test for the testing framework, which
found the error:

```
Traceback (most recent call last):
  File "test/test_dispatch.py", line 164, in test_def_impl_schema_mismatch
    ], raises=True)
  File "test/test_dispatch.py", line 135, in commute
    results=results, raises=raises)
  File "test/test_dispatch.py", line 83, in run_permutation
    .format(ctor_order[:i], op_ix))
  File "test/test_dispatch.py", line 59, in check_invariants
    .format(expected_provenance, actual_provenance)
AssertionError: 'name[16 chars]ema: (none)\ncatchall: boxed unboxed :: (Tenso[18 chars]0)\n' != 'name[16 chars]ema: test::foo(Tensor x, Tensor y) -> (Tensor)[53 chars]0)\n'
  name: test::foo
- schema: (none)
+ schema: test::foo(Tensor x, Tensor y) -> (Tensor)
  catchall: boxed unboxed :: (Tensor _0) -> (Tensor _0)
 : expected from running ctors (1,); actual from running ctors (1,) and then failing to run ctor 0 (did this failure leave the dispatcher in a wedged state? it shouldn't!)
```

Coming soon: direct C++ tests for basic API surface area

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

smessmer commented on test/test_dispatch.py:
```diff
@@ -0,0 +1,359 @@
+    # don't commute, you can still run commute with a fixed ctor_order
+    # so that you can test that the destructors still commute
+    def run_ops(self, name, ops, ctor_order=None, dtor_order=None,
+                results=None, raises=False):
```
nit: rename raises to allow_raises or expect_raises, depending on if passing in True and not raising is an error

ezyang replied:
https://github.com/pytorch/pytorch/pull/35061/commits/cbc194d310c4fc7261986be5bd67cd97983bcfd7

smessmer commented on test/test_dispatch.py:
```diff
@@ -0,0 +1,359 @@
+''')
+
+    # TODO: get rid of this test when multiple defs are wrong
+    def test_multiple_def_schema_mismatch(self):
```
You can probably add a few tests for overload names, i.e. multiple defs are valid if different overload names.

ezyang replied:
I'm going to add this as a C++ test

ezyang replied:
https://github.com/pytorch/pytorch/pull/35061/commits/ab6ab1ebc298f4a40dc808e61025d8f40c15c3d0

smessmer commented on test/test_dispatch.py:
```diff
@@ -0,0 +1,359 @@
+    #
+    # NB: checking all permutations means this function is exponential in
+    # the length of ops!  So don't pass too many ops to this function!
+    def commute(self, name, ops, ctor_order=None, raises=False):
```
Do you think it makes sense to call the registered ops from python and verify their outputs to make sure everything is wired up correctly?

ezyang replied:
I'm relying on dumpState to tell me if the functions are registered correctly. There's some calling tests in C++, but I couldn't figure out a way to conveniently make invocations from Python and setup some sort of mocking API, so I decided not to do it.

ezyang commented on aten/src/ATen/core/op_registration/op_registration.cpp:
```diff
@@ -97,24 +97,29 @@ void RegisterOperators::registerOp_(Options&& options) {
 RegisterOperators::~RegisterOperators() = default;
-RegisterOperators::RegisterOperators(RegisterOperators&&) noexcept = default;
-RegisterOperators& RegisterOperators::operator=(RegisterOperators&&) noexcept = default;
+RegisterOperators::RegisterOperators(RegisterOperators&&) = default;
```
This modification was a mistake and it was fixed in the master PR.

ezyang commented on aten/src/ATen/core/op_registration/op_registration.h:
```diff
@@ -733,6 +739,30 @@ inline CppFunction dispatch_autograd(Func&& raw_f) {
+}
+inline FunctionSchema&& schema(FunctionSchema&& s) { return std::move(s); }
+
+inline c10::either<OperatorName, FunctionSchema> constructSchemaOrName(FunctionSchema&& s) {
```
Yeah, I can dunk this in a detail namespace. My original thinking was that things I exposed in `torch` namespace were the actual public ones, but that's a very subtle distinction.

ezyang replied:
Done in https://github.com/pytorch/pytorch/pull/35061/commits/3955c7182de657cf835f61e684bab4b93de3254c

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -137,5 +139,54 @@ void OperatorEntry::updateDispatchTable_(c10::optional<DispatchKey> dispatch_key
 }
 
+void OperatorEntry::checkInvariants() const {
+  if (schema_) {
```
> Did you evaluate if we can delete the name_ member and just use dispatchTable_.operatorName() everywhere instead?

I suppose this is possible, although that kind of seems like a smell to me; logically, it's the OperatorEntry that has a name, and the DispatchTable is just an internal implementation detail of the actual "fast path" table. So a better refactor would be to see how to eliminate `dispatchTable_.operatorName()` (but I can't conveniently do that because then I can't print good error messages in dispatch table!) So where this actually makes me lean is that DispatchTable should be folded into OperatorEntry. What do you think about this?

ezyang commented on aten/src/ATen/core/boxing/KernelFunction.cpp:
```diff
@@ -16,4 +18,27 @@ void fallthrough_kernel(OperatorKernel*, const OperatorHandle&, Stack*) {
 }
 
+// single line summary of state
+std::string KernelFunction::dumpState() const {
```
I feel like the likelihood of misuse is low (since the names are reasonably clear), and I actually *want* people to use these methods outside of test cases, e.g., when they are debugging. I definitely don't want them to be private; if you're debugging some problem and you want to stick a printf debug, I want to let you do that. If you like, I could add leading underscores to them, how does that sound?

ezyang commented on aten/src/ATen/core/op_registration/op_registration.h:
```diff
@@ -748,75 +778,131 @@ inline CppFunction dispatch_autograd(Func&& raw_f) {
+  Module& def(FunctionSchema&& schema) &;
+  template <typename Schema>
+  Module&& def(Schema&& raw_schema) && {
+    FunctionSchema s = schema(std::forward<Schema>(raw_schema));
```
Oops, that's right!

ezyang replied:
https://github.com/pytorch/pytorch/pull/35061/commits/3955c7182de657cf835f61e684bab4b93de3254c

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -71,8 +76,11 @@ class OperatorEntry final {
+  void deregisterKernel_(c10::optional<DispatchKey> dispatch_key, std::list<KernelEntry>::iterator kernel);
+
+  std::string dumpState() const;
+  void checkInvariants() const;
```
Yeah, I vacillated a little about this. In the end, I added one test to `test/test_dispatch.py` to check if invariants were satisfied for all operators that were registered. There isn't a convenient place to call this function in prod; if you call it at registration time, you will keep testing invariants for every registration, and I imagine it will be noticeable at that point. You definitely want to be able to turn off invariant checking, because in general it might be expensive to do this check.
